libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,404 @@
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void RedundantStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	RedundantStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	// TODO handle reverts of user-defined functions
+
+	if (BuiltinFunction const* f = m_dialect.builtin(_functionCall.functionName.name))
+		if (f->controlFlowSideEffects.terminates)
+		{
+			changeUndecidedTo(State::Unused, Location::Memory);
+			changeUndecidedTo(
+				f->controlFlowSideEffects.reverts ?
+				State::Unused :
+				State::Used,
+				Location::Storage
+			);
+		}
+}
+
+void RedundantStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	RedundantStoreBase::operator()(_functionDefinition);
+}
+
+
+void RedundantStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void RedundantStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	RedundantStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
+		vector<Operation> operations = operationsFromFunctionCall(*funCall);
+		yulAssert(operations.size() == 1, "");
+		m_storeOperations[&_statement] = move(operations.front());
+	}
+}
+
+vector<RedundantStoreEliminator::Operation> RedundantStoreEliminator::operationsFromFunctionCall(
+	FunctionCall const& _functionCall
+) const
+{
+	using evmasm::Instruction;
+
+	YulString functionName = _functionCall.functionName.name;
+	SideEffects sideEffects;
+	if (BuiltinFunction const* f = m_dialect.builtin(functionName))
+		sideEffects = f->sideEffects;
+	else
+		sideEffects = m_functionSideEffects.at(_functionCall.functionName.name);
+
+	if (optional<evmasm::Instruction> instruction = toEVMInstruction(m_dialect, functionName))
+	{
+		switch (*instruction)
+		{
+		case Instruction::SSTORE:
```

Comment:
This list did not include `MLOAD`, but I'm not able to create a test against it...